serialization/src/commonMain/kotlin/games/studiohummingbird/cultoftheancestormoth/serialization/BethesdaBufferDecoder.kt

```diff
@@ -0,0 +1,123 @@
+/**
+Cult of the Ancestor Moth (BethesdaBufferDecoder.kt)
+Copyright (C) 2024  Zymus (moore.zyle@gmail.com)
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published
+by the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package games.studiohummingbird.cultoftheancestormoth.serialization
+
+import games.studiohummingbird.cultoftheancestormoth.serialization.datatypes.InlineNullTerminatedString
+import games.studiohummingbird.cultoftheancestormoth.serialization.datatypes.nullTerminatedStringDecoder
+import kotlinx.io.Source
+import kotlinx.io.readByteArray
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.AbstractDecoder
+import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.encoding.CompositeDecoder.Companion.DECODE_DONE
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+
+@ExperimentalStdlibApi
+@ExperimentalSerializationApi
+class BethesdaBufferDecoder(private val source: Source) : AbstractDecoder() {
+
+    override val serializersModule: SerializersModule = EmptySerializersModule()
+
+    private val primitiveBufferDecoder by lazy { PrimitiveBufferDecoder(source) }
+
+    private var structureElements = 0
+    private var structureElementIndex = 0
+
+    override fun beginStructure(descriptor: SerialDescriptor): CompositeDecoder {
+        println("beginStructure")
+        println("- kind=${descriptor.kind}")
+        println("- serialName=${descriptor.serialName}")
+        println(". isRecord=${descriptor.isRecord()}")
+        println("- annotations=${descriptor.annotations}")
+        println("- elementsCount=${descriptor.elementsCount}")
+        repeat(descriptor.elementsCount) {
+            println(" - elementDescriptor$it=${descriptor.getElementDescriptor(it)}")
+        }
+        return BethesdaBufferDecoder(source).also {
+            it.structureElements = descriptor.elementsCount
+            it.structureElementIndex = 0
+        }
+    }
+
+    override fun endStructure(descriptor: SerialDescriptor) {
+        structureElements = 0
+        structureElementIndex = 0
+    }
+
+    override fun decodeByte(): Byte {
```

Comment:
If we remove the explicit logs and explicit variable, we could condense these primitive ones to
```suggestion
    override fun decodeByte(): Byte = primitiveBufferDecoder.decodeByte()
```